Magnetic Field Propagator (#49)

whokion opened the pull request:
The first version of FieldPropagator and associated tests:
- Added FieldTrackView
- Added FieldPropagator
- Added state accessors of GeoTrackView to support FieldTrackView
- Added FieldPropagator.test

sethrj commented on src/field/FieldTrackView.hh:
```diff
@@ -0,0 +1,95 @@
+                                         const ParticleTrackView& pt_view);
+
+    // Copy Construct
+    inline CELER_FUNCTION FieldTrackView(const FieldTrackView& other) = default;
```
This shouldn't be necessary; a force-defaulted copy constructor should only ever be require if you define a custom move constructor.

sethrj reviewed (changes requested):
I might not be able to resume the review today, so here are a couple of comments. I think we should try to encapsulate all the navstate-related aspects of the `FieldState` into the `GeoTrackView` itself, but that's just my first instinct rather than something I've really thought through about what parts (e.g. step and safety) belong where.

whokion commented on src/field/FieldPropagator.i.hh:
```diff
@@ -0,0 +1,198 @@
+ */
+CELER_FUNCTION real_type FieldPropagator::operator()(FieldTrackView* view)
+{
+    CELER_ASSERT(view->charge().value() != 0);
```
Done

whokion commented on src/field/FieldPropagator.i.hh:
```diff
@@ -0,0 +1,198 @@
+        // Add sub-step until there is no remaining step length
+        step_taken += sub_step;
+
+    } while (!intersect.status && (step_taken + shared_.minimum_step) < step);
```
Changed to intersected.

whokion commented on src/field/FieldPropagator.hh:
```diff
@@ -0,0 +1,80 @@
+
+  private:
+    // A helper input/output for private member functions
+    struct IntersectionIO
```
Yes, I meant IO as "in and out", but agree that it is confusing.  Changed to Intersection.

whokion commented on src/field/FieldPropagator.i.hh:
```diff
@@ -0,0 +1,198 @@
+ */
+CELER_FUNCTION
+void FieldPropagator::check_intersection(FieldTrackView* view,
+                                         const Real3     bed_pos,
```
Good catch!  Fixed.

whokion commented on src/field/FieldTrackView.hh:
```diff
@@ -0,0 +1,95 @@
+    CELER_FUNCTION bool& on_boundary() { return on_boundary_; }
+    CELER_FUNCTION real_type& step() { return step_; }
+    CELER_FUNCTION OdeState& state() { return state_; }
+    CELER_FUNCTION real_type& safety() { return safety_; }
```
Replaced!

whokion commented on src/field/FieldUtils.i.hh:
```diff
@@ -67,7 +67,7 @@ CELER_FUNCTION real_type distance_chord(const OdeState& beg_state,
 
     Real3 cross = cross_product(beg_end, beg_mid);
-    return std::sqrt(dot_product(cross, cross) / dot_product(beg_end, beg_end));
+    return norm(cross) / dot_product(beg_end, beg_end);
```
Oops!  It should be norm(cross)/norm(beg_end).  Reverted to the previous expression as it uses one sqrt instead of two when using norm.

whokion commented on src/field/FieldPropagator.i.hh:
```diff
@@ -0,0 +1,198 @@
+
+        // Adjust the intersect scale in the case that it is unity
+        if (intersect->scale == 1)
+            intersect->scale *= FieldPropagator::scale_factor();
```
Yes, the correct logic should be
<pre>
intersect->intersected = (linear_step <= length);
</pre>
which should be enough without the fuzzy scaling factor.  However, It fails to find an intersection without this fuzzy scaling (looping) when linear_step == length, which I do not understand yet.  I will investigate the case more carefully and try to come up with the better logic later.

whokion replied:
I removed this fuzzy scaling factor now and improved the logic for finding the intersection point in general (including other things related to the boundary crossing workflow).

pcanal commented on src/field/FieldPropagator.i.hh:
```diff
@@ -0,0 +1,199 @@
+{
+    intersect->intersected = false;
+
+    Real3 chord = end_pos;
```
I am not sure but would re-using the argument mem space/register be more efficient (i.e `const Real3 end_pos` -> `Real3 end_pos` and maybe alias it (via a ref) for code readability)?

whokion replied:
I guess that both **Real3 end_pos** and **const Real3& end_pos** are acceptable, but not **Real3& end_pos.** as it is mutable.  For now, I change it to const Real3& end_pos.

whokion commented:
@sethrj Regarding to navstate-related interfaces, I am totally open to move things to GeoTrackView, which potentially remove  unnecessary public accessors of vgstates which I introduced.

sethrj commented on src/field/FieldPropagator.hh:
```diff
@@ -0,0 +1,79 @@
+    FieldPropagator(const FieldParamsPointers& shared, FieldDriver& driver);
+
+    // Propagation in a field
+    inline CELER_FUNCTION real_type operator()(FieldTrackView* view);
```
I think this highest-level class should look like the `LinearPropagator`, where the track view is part of the class constructor, and the operators take the physical step limit (or an option to move to the next geometry boundary with an "infinite" physical step).

whokion replied:
Contrary to the LinearPropagator, this operator also need information from ParticleTrackView (i.e, momemtum, charge, (global time, spin later)) which is why this intermediate FieldTrackView was originally introduced - the scope of its usage is now limited up to the high-level, but may be extended deeper later.   Again, the difference between LinearPropagator and the FieldPropagator is whether the (propagation) state (not necessarily the vecgeom navigatiion state for the field propagation) is occasionally updated or not.  I guess that it is up to the transport process what relevant information passes to the linear propagator for the neutral particle  and to the field propagator for the charged particle with a non-zero magnetic field.  Anyway, we can definitely discuss these infra-structure related consistency further and remove or move-round things later once the current implementation is efficiently working for the case with non-uniform magnetic field (which is the next target).

sethrj reviewed (changes requested):
OK, I think the questions about  the persistent state (safety distance) and `FieldPropagator` will help shape how we can move some of the pieces into the GeoTrackView.

whokion commented on src/field/FieldPropagator.hh:
```diff
@@ -0,0 +1,79 @@
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Brief class description.
```
Added more description.

whokion commented on src/field/FieldTrackView.i.hh:
```diff
@@ -0,0 +1,63 @@
+    , vgnext_(geo_view.vgnext())
+{
+    state_.pos = geo_view.pos();
+    axpy(particle_view.momentum().value(), geo_view.dir(), &state_.mom);
```
Added the typedef.  It would be good to have something like  "scale" (?).

whokion commented on src/field/FieldTrackView.i.hh:
```diff
@@ -0,0 +1,63 @@
+
+namespace celeritas
+{
+// Construct from GeoTrackView and ParticleTrackView
```
Fixed the format.

whokion commented on src/field/FieldTrackView.hh:
```diff
@@ -0,0 +1,104 @@
+    CELER_FUNCTION bool on_boundary() const { return on_boundary_; }
+    CELER_FUNCTION real_type step() const { return step_; }
+    CELER_FUNCTION const OdeState& state() const { return state_; }
+    CELER_FUNCTION real_type safety() const { return safety_; }
+    CELER_FUNCTION Real3 origin() const { return origin_; }
```
This can be discussed further even though this safety origin may not be used anywhere else  outside the field propagation in which the step may be sub-divided, if necessary, and the track is moved to an intermediate state where the safety is query to ensure whether it is inside the volume or not. If so, the safety origin is updated for the next trial stub-step within the current volume and so on.   In the linear navigator, the origin is the current position and never get updated within the current volume.

whokion commented:
@sethrj I guess that I resolved most of outstanding issues for this MR besides those related to GeoTrackView and utility interfaces/functions using the vecgeom navigator, which we may be addressed in a separate MR.   Can review the change
and consider to merge this MR if there are no obvious things to changes at this moment.  Thanks.

sethrj commented:
@whokion Thanks, sorry I didn't have a chance to merge this or review it in more detail. Can you, me, @pcanal, @amandalund, maybe @tmdelellis interested meet today at 2 pm eastern/3 central to discuss the interface? I'd like to get that  in place before resuming the review.

whokion commented:
@sethrj Yes, I am available anytime this afternoon.

whokion commented:
@sethrj I would like to close this MR (delete the branch) and open a new one as this branch adding new files has been significant changed from the initial version (i.e, removed FieldTrackView and un-needed public (vgstate) accessors of GeoTrackView which were introduced for the defunct FieldTrackView, and changed many interfaces with using GeoTracView instead of FieldTrackView and etc.) - you may take a quick look at the current branch, which is very close to an upcoming new one except a coupe of more clean ups (such as returning pos and dir instead of OdeState which should be agnostic to GeoTrackView or the higher level transport interface or class.  One more question is whether FieldPropagator can modify GeoTrackView data (i.e, pos and dir by available stters) directly.  In the real work flow, I think it should be the case, but we may also send an intermediate result of the field propagation to the high level transportation manager which can update things accordingly in one central place.  As shortly discussed last Wednesday, we agreed to go with the latter (which is reflect in the current implementation), but the first option will also reduce unnecessary data copy.  Either is okay for me now, but just want to know your preference.  One more note is that the current field propagator updates the final vgstate inside the class anyway as the final direction from the last sub-step before the boundary is used for the final direction for updating the vgstate, which is different from the final (momentum) direction by the stepper.  Anyway, please let me know whether I can delete this one and create a new one.  Thanks.

whokion commented:
@sethrj This MR is now ready to be reviewed.  Thanks.

sethrj commented on test/CMakeLists.txt:
```diff
@@ -243,5 +243,28 @@ celeritas_setup_tests(SERIAL PREFIX field)
+
+  if(CELERITAS_USE_CUDA)
+    add_library(CeleritasPropagatorTest field/FieldPropagator.test.cu)
+    target_link_libraries(CeleritasPropagatorTest
+      PRIVATE 
+      celeritas
+      VecGeom::vecgeomcuda
+      VecGeom::vecgeomcuda_static
+    )
+    set_target_properties(CeleritasPropagatorTest PROPERTIES
+      LINKER_LANGUAGE CUDA
+      CUDA_SEPARABLE_COMPILATION ON
+      POSITION_INDEPENDENT_CODE ON
+    )
```
I added a CMake macro at some point to hide the extra flags (and so we can change them all in one place if needed). I  think this should work:

```suggestion
    target_link_libraries(CeleritasPropagatorTest
      PRIVATE celeritas
    )
    celeritas_link_vecgeom_cuda(CeleritasPropagatorTest)
```

whokion replied:
Thanks for a simplification!  Modified accordingly.

sethrj reviewed (approved):
I'm satisfied. Thanks for the iterations!

whokion commented on src/field/FieldDriver.hh:
```diff
@@ -77,6 +77,21 @@ class FieldDriver
+
+    inline CELER_FUNCTION real_type max_nsteps() { return shared_.max_nsteps; }
+
+    inline CELER_FUNCTION real_type delta_intersection()
```
Added missing const qualifier.

whokion commented on src/field/FieldPropagator.hh:
```diff
@@ -0,0 +1,89 @@
+ * Brief class description.
+ *
+ *  This is a high level interface of propagating a charged particle in a
+ *  magnetic field. For a given initial state (position, momentum), it
```
Fixed.

sethrj commented:
Thanks @whokion!

whokion commented:
@sethrj Thanks a lot for all reviews and the merge.  The next plan is to add/support 1) a non-uniform magnetic field case, 2) a small angle approximation for detecting the boundary intersection.  Of course, I will closely follow up related interfaces to geometry/transport processes as well as further improvement/optimization of existing field codes.  If you have a different priority and plan, please let's discuss at our usual meetings.
